### ISSUE TITLE
Train_loss and test_loss become NaN after using model.prune()

### ISSUE THREAD

**Issue by DonSteven** [KindXiaoming/pykan#416]
Hello,
I'm working on a multi-classification model and the result looks good. But when I train the model after using model.prune(), train_loss and test_loss become NaN easily even if I set lr and step very small. Could anyone please to solve out this problem? Thanks a lot!

Here's my code. I include a scheduler: torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=T_max, eta_min=eta_min)

![image](https://github.com/user-attachments/assets/cc9f7ef4-5613-4f3e-bca5-f92ec6754627)
![image](https://github.com/user-attachments/assets/dfce3839-abdc-474b-9a38-13caa8a180d8)
![image](https://github.com/user-attachments/assets/c533e35b-7f70-46d6-9b47-fd5f25b328c2)
![image](https://github.com/user-attachments/assets/3cad0dd4-7c66-476b-b527-137ce54a3e48)
![image](https://github.com/user-attachments/assets/cbf69ddd-fce3-40fa-8614-3d18ed51439e)

**Comment by shrrr** [comment]
I have the same problem. I also use KANs to do five classification tasks, but after I complete the first training stage and run ```model = model.prune()```, I get **nan** in train loss and test loss in the second training stage. However, after I set ```update_grid=False``` in ```model.fit()``` function, the loss is not **nan**. I don't know why!!

The model I defined is:
```
model = KAN(width=[9,10,2], grid=3, k=3, seed=42, device=device)
```
And training code is:
```
results = model.fit(data, opt="LBFGS", metrics=(train_acc, test_acc), loss_fn=torch.nn.CrossEntropyLoss(), steps=50)
```

Any helpful information or ideas, please let me know! Thank you!

**Comment by DonSteven** [comment]
> I have the same problem. I also use KANs to do five classification tasks, but after I complete the first training stage and run `model = model.prune()`, I get **nan** in train loss and test loss in the second training stage. However, after I set `update_grid=False` in `model.fit()` function, the loss is not **nan**. I don't know why!!
> 
> The model I defined is:
> 
> ```
> model = KAN(width=[9,10,2], grid=3, k=3, seed=42, device=device)
> ```
> 
> And training code is:
> 
> ```
> results = model.fit(data, opt="LBFGS", metrics=(train_acc, test_acc), loss_fn=torch.nn.CrossEntropyLoss(), steps=50)
> ```
> 
> Any helpful information or ideas, please let me know! Thank you!

Thanks for your reply! It really works！

**Comment by andrewrgarcia** [comment]
@DonSteven might not be the best fix but I had the same problem, so I have reverted to `pykan==0.2.2` (where this issue is no longer present) until resolved in the most updated version

**Comment by srigas** [comment]
Based on my very short experience working with the new version of `pykan`, I have noticed that the issue arises whenever disconnected nodes remain at the KAN's computation graph after pruning. The plot provided in your post appears to fall under this case. My workaround was to define a custom prune function, as follows:

```
def custom_prune(model, node_th=1e-2, edge_th=3e-2):
    # First prune edges
    model.prune_edge(edge_th, log_history=False)
    model.forward(model.cache_data)
    model.attribute()
    model.log_history('prune')
    # Then prune nodes
    model = model.prune_node(node_th, log_history=False)
    
    return model

# Example usage
model = custom_prune(model)
```

Essentially, all I've done is reverse the pruning of edges and pruning of nodes, ensuring that edges are pruned **before** nodes and not vice-versa, as in the original `prune()` implementation. This appears to fix the issue of NaNs post-pruning, although I'm not sure if it affects the training in some other, unexpected way.

**Comment by KindXiaoming** [comment]
@srigas this is an interesting observation! it's be awesome if ppl can confirm this for their cases!

**Comment by Ilex00para** [comment]
I encountered the same problem and it worked with @srigas function and also when you change the order of node and edge pruning in the library function. 

When using the standard `.prune()` function, it seems that the pruning changes are made in the mask (adding zeros) but not in the attribution score calculations (since they do not change when using the normal pruning function but the costume function). Also, if the `.prune_edge()` is not called no pruning is done at all even after calling `.prune_node()`.
As I understood from the code the node pruning depends (only) on the edges and the functions are called in the wrong order? 

(Also some things for the documentation prune_edge is not returning a new KAN its just updating the mask.)

**Comment by joshua-cogliati-inl** [comment]
For what it is worth, here is some more example code that triggers the NaN after model.prune() in case anyone needs to test a fix.
```python
import kan

kan.torch.set_default_dtype(kan.torch.float64)
device = kan.torch.device('cuda' if kan.torch.cuda.is_available() else 'cpu\
')
print(device)

model = kan.KAN(width=[2,5,1], grid=3, k=3, seed=42, device=device)
import kan.utils

f = lambda x: x[:,[0]]**3+x[:,[1]]**2

dataset = kan.utils.create_dataset(f, n_var=2, device=device)
print(dataset['train_input'].shape, dataset['train_label'].shape)

model(dataset['train_input']);
import matplotlib.pyplot as plt
#model.plot()
#plt.show()

model.fit(dataset, opt="LBFGS", steps=50, lamb=0.001)
#model.plot()
#plt.show()

for i in range(2):
    model.fit(dataset, opt="LBFGS", steps=50)
    model.fit(dataset, opt="LBFGS", steps=50)
    print(i)
    #model.plot()
    #plt.show()


model = model.prune()
model.plot()
plt.show()


model.fit(dataset, opt="LBFGS", steps=50)
```

**Comment by joshua-cogliati-inl** [comment]
I continue to try and learn a bit more about it.  Whatever prune does still allows forward pass to work, but not the fit.  I am suspicious of `optimizer.step(closure) ` in `fit` but am not sure if that is where things start generating `NaN`s.
```python
import kan
import torch

kan.torch.set_default_dtype(kan.torch.float64)
device = kan.torch.device('cuda' if kan.torch.cuda.is_available() else 'cpu\
')
print(device)

model = kan.KAN(width=[2,5,1], grid=3, k=3, seed=42, device=device)
import kan.utils

f = lambda x: x[:,[0]]**3+x[:,[1]]**2

dataset = kan.utils.create_dataset(f, n_var=2, device=device)
print(dataset['train_input'].shape, dataset['train_label'].shape)

model(dataset['train_input']);
import matplotlib.pyplot as plt
#model.plot()
#plt.show()

model.fit(dataset, opt="LBFGS", steps=50, lamb=0.001)
#model.plot()
#plt.show()

for i in range(2):
    model.fit(dataset, opt="LBFGS", steps=50)
    model.fit(dataset, opt="LBFGS", steps=50)
    print(i)
    #model.plot()
    #plt.show()


def custom_prune(model, node_th=1e-2, edge_th=3e-2):
    # First prune edges
    model.prune_edge(edge_th, log_history=False)
    model.forward(model.cache_data)
    model.attribute()
    model.log_history('prune')
    # Then prune nodes
    model = model.prune_node(node_th, log_history=False)
    
    return model

def show_model(model):
    print(f"node_bias: {model.node_bias}")
    print(f"node_scale: {model.node_scale}")
    print(f"subnode_bias: {model.subnode_bias}")
    print(f"subnode_scale: {model.subnode_scale}")

model2 = model.copy()
model2 = model2.prune()
#model.plot()
#plt.show()
#model2.plot()
#plt.show()

sample_input = torch.randn((1, model.width[0][0])).to(model.device)
model_output = model.forward(sample_input)
model2_output = model2.forward(sample_input)
print("before output",model_output)
print("after_output",model2_output)

loss = model2_output.sum()
loss.backward()
for name, param in model2.named_parameters():
    if param.grad is not None:
        print(f"Gradient for {name}:", param.grad)

breakpoint()
#This fit will result in lots of nan's in the gradient
model2.fit(dataset, opt="LBFGS", steps=1)


for name, param in model2.named_parameters():
    if param.grad is not None:
        print(f"Gradient for {name}:", param.grad)

```
